realm/realm-library/src/main/java/io/realm/ProxyState.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2016 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Future;
+
+import io.realm.internal.RealmObjectProxy;
+import io.realm.internal.Row;
+import io.realm.internal.Table;
+import io.realm.internal.TableQuery;
+import io.realm.internal.log.RealmLog;
+
+/**
+ * This implements {@code RealmObjectProxy} interface, to eliminate copying logic between
+ * {@link RealmObject} and {@link DynamicRealmObject}.
+ */
+public final class ProxyState implements RealmObjectProxy {
```

Comment:
please note that latest changes uses `ProxyState` instance inside  `RealmObjectProxy` (to reduce the number of method the proxy need to implement) we need  `ProxyState` since it's referenced from an internal package